sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,934 @@
+        [kDifUsbdevIrqRxBitstuffError] = USBDEV_INTR_STATE_RX_BITSTUFF_ERR,
+        [kDifUsbdevIrqFrame] = USBDEV_INTR_STATE_FRAME,
+        [kDifUsbdevIrqConnected] = USBDEV_INTR_STATE_CONNECTED,
+};
+
+static bool is_valid_irq(dif_usbdev_irq_t irq) {
+  return irq >= kDifUsbdevIrqFirst && irq <= kDifUsbdevIrqLast;
+}
+
+/**
+ * Static functions for the free buffer pool.
+ */
+
+static bool buffer_pool_is_full(dif_usbdev_buffer_pool_t *pool) {
+  return pool->top == (USBDEV_NUM_BUFFERS - 1);
+}
+
+static bool buffer_pool_is_empty(dif_usbdev_buffer_pool_t *pool) {
+  return pool->top == -1;
+}
+
+static bool buffer_pool_is_valid_entry(uint8_t entry) {
+  return entry < USBDEV_NUM_BUFFERS;
+}
+
+static bool buffer_pool_add(dif_usbdev_buffer_pool_t *pool, uint8_t entry) {
+  if (buffer_pool_is_full(pool) || !buffer_pool_is_valid_entry(entry)) {
+    return false;
+  }
+
+  pool->entries[++pool->top] = entry;
+
+  return true;
+}
+
+static bool buffer_pool_init(dif_usbdev_buffer_pool_t *pool) {
+  // Start with an empty pool
+  pool->top = -1;
+
+  // Add all entries
+  for (uint8_t i = 0; i < USBDEV_NUM_BUFFERS; ++i) {
+    if (!buffer_pool_add(pool, i)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static bool buffer_pool_remove(dif_usbdev_buffer_pool_t *pool,
+                               uint32_t *entry) {
+  if (buffer_pool_is_empty(pool) || entry == NULL) {
+    return false;
+  }
+
+  *entry = pool->entries[pool->top--];
+
+  return true;
+}
+
+/**
+ * Static functions for managing read and write buffer operations.
+ *
+ * In `dif_usbdev_t.active_buffer_ops`:
+ * - Bits 0..11: True if writing bytes of an outgoing packet for endpoint N
+ * - (0..11).
+ * - Bits 12..30: Unused.
+ * - Bit 31: True if reading bytes of an incoming packet.
+ *
+ * Note: The first parameters of the functions below are `dif_usbdev_t *` in an
+ * attempt to avoid confusion in cases where the second parameter is a
+ * `uint32_t`.
+ */
+
+static void buffer_op_init(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops = 0;
+}
+
+static void buffer_op_start_write(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  usbdev->active_buffer_ops |= (1 << kEndpointHwInfos[endpoint].bit_index);
+}
+
+static void buffer_op_stop_write(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  usbdev->active_buffer_ops &= ~(1 << kEndpointHwInfos[endpoint].bit_index);
+}
+
+static bool buffer_op_is_writing(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  return (usbdev->active_buffer_ops &
+          (1 << kEndpointHwInfos[endpoint].bit_index)) > 0;
+}
+
+static void buffer_op_start_read(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops |= (1 << USBDEV_READ_BUFFER_OP_BIT_INDEX);
+}
+
+static void buffer_op_stop_read(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops &= ~(1 << USBDEV_READ_BUFFER_OP_BIT_INDEX);
+}
+
+static bool buffer_op_is_reading(dif_usbdev_t *usbdev) {
+  return (usbdev->active_buffer_ops & (1 << USBDEV_READ_BUFFER_OP_BIT_INDEX)) >
+         0;
+}
```

Comment:
I think I'm eating my words from https://github.com/lowRISC/opentitan/pull/2047, we really aught to have non-MMIO get_bit and set_bit functions. I'll open a CL to add them, and you'll be able to use them here once I'm done.

@silvestrst You were right. =P

Reply:
Sure, I don't think we had those when I was writing this. Done.